src/plugins/discover/public/components/discover_grid/discover_grid.tsx

```diff
@@ -416,14 +425,45 @@ export const DiscoverGrid = ({
     [dataView, onFieldEdited, services.dataViewFieldEditor]
   );
 
+  const visibleColumns = useMemo(
+    () => getVisibleColumns(displayedColumns, dataView, showTimeCol) as string[],
+    [dataView, displayedColumns, showTimeCol]
+  );
+
+  const cellActionProps = useMemo<UseDataGridColumnsCellActionsProps>(() => {
+    const fields = cellActionsTriggerId
+      ? visibleColumns.map((columnName) => {
+          const field = dataView.getFieldByName(columnName)?.spec;
+          return {
+            name: columnName,
+            type: field?.type ?? 'keyword',
+            values: displayedRows.map((row) => {
+              const value = row.flattened?.[columnName];
+              const valuesArray = Array.isArray(value) ? value : [value];
+              return valuesArray;
+            }),
```

Comment:
The `useDataGridColumnsCellActions` hook returns the `EuiDataGridColumnCellAction[]` for each field, this is the type that `EuiDataGrid` needs for the column's `cellActions` property, which is an array of functions that only receive `rowIndex` to identify what are the values that we have to use for that specific cell action:

https://github.com/elastic/kibana/blob/76f6fd9ef330c2be34d8151bbccc7b29afe63f92/packages/kbn-cell-actions/src/hooks/use_data_grid_column_cell_actions.tsx#L100

So we need all the row values of each column when we build the whole structure of `EuiDataGridColumnCellAction` functions, in order to extract the proper value from the row index when the action is executed.

Do you think there's any scenario where this could be a problem?

Reply:
Yes, basically the default scenario. Discover should be able to handle large number of rows and columns. Also we are planning to integrate pagination with lazy loading - so number of rows will only grow.

My suggestion would be to update `useDataGridColumnsCellActions` so it can accept a function which would give a value by `rowIndex` and field name in return. Quick example of it:
```
const getCellValue = useCallback((fieldName, rowIndex) => {
 return displayedRows[rowIndex].flattened[fieldName]
}, [displayedRows])

const columnsCellActions = useDataGridColumnsCellActions(cellActionProps, getCellValue);
```

The idea is to avoid unnecessary iterations over fields array or rows array. Could you please double check `fields.map` calls in `useDataGridColumnsCellActions` too? Can it be optimized?

Since EuiGrid supports virtualization and it does not render all data at the same time, we should try to avoid extra calculations over all data too.